model/fn-execution/src/main/proto/beam_provision_api.proto

```diff
@@ -67,6 +67,10 @@ message ProvisionInfo {
     // (optional) Resource limits that the SDK harness worker should respect.
     // Runners may -- but are not required to -- enforce any limits provided.
     Resources resource_limits = 4;
+
+    // (required) The artifact retrieval token produced by
+    // ArtifactStagingService.CommitManifestResponse.
+    string retrieval_token = 6;
```

Comment:
This design was discussed some time ago (see also https://github.com/apache/beam/pull/5582) - we're coming from the assumption that all services are (or at least may be) globally distributed and stateless, i.e. we're not relying on the assumption that there's 1 ArtifactRetrievalService per worker or per harness. Without that assumption, we need the ArtifactRetrievalService calls to be somehow linked to which job we're talking about. Likewise, ArtifactStagingService also needs to know which job we're talking about.

We decided to do this by propagating tokens:

- PrepareJob returns a token used for ArtifactStagingService calls
- ArtifactStagingService.CommitManifest returns a token used for ArtifactRetrievalService calls

This token is an opaque string containing the information necessary for the service to do its job. In practice, with the "distributed file system" based implementations of both services, we're using (basically) a base path as the token.

Alternatively we could explicitly include the job ID in the RPCs, but that would require the services to do some sort of global lookup of artifact placement parameters based on job ID, it seems easier to include the necessary parameters explicitly in the token.

Now, since a retrieval token is needed for the harness to talk to the retrieval service, it seems reasonable to include it in provision info. It's not part of the service descriptor because it does not identify the service, it only gives a necessary argument for its RPCs.

Reply:
The original design has the same assumptions. The provisioning and control service will still have to differentiate harnesses based on the client id (using internal lookups and whatnot). Artifact service is now different. 

For the artifact ApiServiceDescriptor, I meant that if we must call provisioning first for the token then it might as well vendor the service descriptor, too (and perhaps the control service as well towards a smaller container contract). That would be a larger change, though.

Reply:
Ack, we haven't thought about applying similar changes to provision and control services but it makes sense too - might do it later if we run into issues stemming from that.